src/titan_checkpoint_impl.cc

```diff
@@ -0,0 +1,292 @@
+      } else {
+        s = db_->GetEnv()->DeleteFile(subchild_path);
+        ROCKS_LOG_INFO(info_log, "Delete file %s -- %s", subchild_path.c_str(),
+                       s.ToString().c_str());
+      }
+    }
+  }
+  // Finally delete the private dir
+  s = db_->GetEnv()->DeleteDir(full_private_path);
+  ROCKS_LOG_INFO(info_log, "Delete dir %s -- %s",
+                 full_private_path.c_str(), s.ToString().c_str());
+}
+
+Status TitanCheckpointImpl::CreateTitanManifest(const std::string& file_name,
+                                                std::vector<VersionEdit>* edits) {
+  Status s;
+  Env* env = db_->GetEnv();
+  bool use_fsync = db_->GetDBOptions().use_fsync;
+  const EnvOptions env_options;
+  std::unique_ptr<WritableFileWriter> file;
+  
+  {
+    std::unique_ptr<WritableFile> f;
+    s = env->NewWritableFile(file_name, &f, env_options);
+    if (!s.ok()) return s;
+    file.reset(new WritableFileWriter(std::move(f), file_name, env_options));
+  }
+  std::unique_ptr<log::Writer> manifest;
+  manifest.reset(new log::Writer(std::move(file), 0, false));
+  
+  for (auto& edit : *edits) {
+    std::string record;
+    edit.EncodeTo(&record);
+    s = manifest->AddRecord(record);
+    if (!s.ok()) return s;
+  }
+  
+  return manifest->file()->Sync(use_fsync);
+}
+
+// Builds an openable checkpoint of TitanDB
+Status TitanCheckpointImpl::CreateCheckpoint(const std::string& checkpoint_dir,
+                                             uint64_t log_size_for_flush) {
+  DBOptions db_options = db_->GetDBOptions();
+  Status s = db_->GetEnv()->FileExists(checkpoint_dir);
+  if (s.ok()) {
+    return Status::InvalidArgument("Directory exists");
+  } else if (!s.IsNotFound()) {
+    assert(s.IsIOError());
+    return s;
+  }
+
+  ROCKS_LOG_INFO(
+          db_options.info_log,
+          "Started the checkpoint process -- creating checkpoint in directory %s",
+          checkpoint_dir.c_str());
+
+  size_t final_nonslash_idx = checkpoint_dir.find_last_not_of('/');
+  if (final_nonslash_idx == std::string::npos) {
+    // npos means it's only slashes or empty. Non-empty means it's the root
+    // directory, but it shouldn't be because we verified above the directory
+    // doesn't exist.
+    assert(checkpoint_dir.empty());
+    return Status::InvalidArgument("Invalid checkpoint directory name");
+  }
+
+  std::string full_private_path =
+          checkpoint_dir.substr(0, final_nonslash_idx + 1) + ".tmp";
+  ROCKS_LOG_INFO(
+          db_options.info_log,
+          "Checkpoint process -- using temporary directory %s",
+          full_private_path.c_str());
+  CleanStagingDirectory(full_private_path, db_options.info_log.get());
+  // Create checkpoint directory and subdirectory
+  s = db_->GetEnv()->CreateDir(full_private_path);
+  if (s.ok()) {
+    s = db_->GetEnv()->CreateDir(full_private_path + "/titandb");
+  }
+  uint64_t sequence_number = 0;
+  if (s.ok()) {
+    // Disable file deletions
+    s = db_->DisableFileDeletions();
```

Comment:
don't need it anymore

Reply:
If we remove `DisableFileDeletions` and `EnableFileDeletions` in CreateCheckpoint , we should move `EnableFileDeletions ` to the last in `CreateCustomCheckpoint`, which will prevent delete files when link or copy files. If we do it in this way, we're essentially moving `DisableFileDeletions` and `EnableFileDeletions` into `CreateCheckpoint `.